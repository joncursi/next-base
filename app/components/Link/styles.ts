/**
 * @prettier
 */

import css from 'styled-jsx/css';

import COLORS from '~/constants/colors';

export default css`
  a {
    color: inherit;
    text-decoration: none;
  }
  a.textLink {
    color: ${COLORS.LINK};
  }
  a.textLink:hover {
    text-decoration: underline;
  }
`;
