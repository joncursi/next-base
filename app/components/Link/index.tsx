/**
 * @prettier
 */

import * as React from 'react';
import isURL from 'validator/lib/isURL';

import Routes from '../../../server/routes';

import styles from './styles';

export interface Props {
  children?: React.ReactNode;
  download?: string | boolean;
  onClick?: (e: React.MouseEvent<HTMLElement>) => void;
  params?: Record<string, any>;
  style?: React.CSSProperties;
  target?: string;
  to: string;
  variant?: 'text';
}

const Link: React.FC<Props> = ({
  children,
  download,
  onClick,
  params,
  style,
  target,
  to,
  variant,
  ...props
}: Props): React.ReactElement<Props> => {
  const isExternalUrl = isURL(to);
  const className = variant === 'text' ? 'textLink' : '';

  return (
    <>
      <style jsx>{styles}</style>

      {isExternalUrl && (
        <a
          className={className}
          download={download}
          href={to}
          onClick={onClick}
          style={style}
          target={target}
          {...props}
        >
          {children || null}
        </a>
      )}

      {!isExternalUrl && (
        <Routes.Link params={params} route={to} {...props}>
          {/* eslint-disable jsx-a11y/no-static-element-interactions */}
          {/* eslint-disable jsx-a11y/click-events-have-key-events */}
          {/* eslint-disable jsx-a11y/anchor-is-valid */}
          <a
            className={className}
            onClick={onClick}
            style={style}
            target={target}
          >
            {children || null}
          </a>
          {/* eslint-enable jsx-a11y/anchor-is-valid */}
          {/* eslint-enable jsx-a11y/click-events-have-key-events */}
          {/* eslint-enable jsx-a11y/no-static-element-interactions */}
        </Routes.Link>
      )}
    </>
  );
};

export default Link;
