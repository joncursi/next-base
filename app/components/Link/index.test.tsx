/**
 * @prettier
 */

import * as React from 'react';
import { noop } from 'lodash';
import renderer from 'react-test-renderer';

import Link from '.';

describe('Link', () => {
  it('matches the snapshot', () => {
    const tree = renderer
      .create(
        <Link
          onClick={noop}
          style={{ color: 'red' }}
          target="_blank"
          to="https://www.google.com"
          variant="text"
        >
          <div>children</div>
        </Link>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
