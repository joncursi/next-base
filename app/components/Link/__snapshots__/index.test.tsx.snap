// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link matches the snapshot 1`] = `
<a
  className="jsx-2857092000 textLink"
  href="https://www.google.com"
  onClick={[Function]}
  style={
    Object {
      "color": "red",
    }
  }
  target="_blank"
>
  <div>
    children
  </div>
</a>
`;
